File: platform/fastapi_server/routes/healthz_elasticsearch_data.route.py
from fastapi import FastAPI, Query, Request
from fastapi.responses import JSONResponse
from db_connection_elasticsearch import ElasticsearchConfig, get_elasticsearch_client


def mount(app: FastAPI):
    """
    Mount Elasticsearch data-exploration routes.
    Read-only endpoints for browsing indices, documents, and field mappings.
    """

    @app.get("/healthz/elasticsearch/indices")
    async def elasticsearch_indices(request: Request):
        client = None
        try:
            config = ElasticsearchConfig()
            client = await get_elasticsearch_client(config)
            raw = await client.cat.indices(format="json")

            indices = [
                {
                    "index": r.get("index"),
                    "health": r.get("health"),
                    "status": r.get("status"),
                    "docsCount": r.get("docs.count"),
                    "storeSize": r.get("store.size"),
                }
                for r in raw
            ]

            return {
                "configuredIndex": config.index or None,
                "indices": indices,
            }
        except Exception as e:
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.close()

    @app.get("/healthz/elasticsearch/indices/{index_name}/mappings")
    async def elasticsearch_index_mappings(index_name: str, request: Request):
        client = None
        try:
            config = ElasticsearchConfig()
            client = await get_elasticsearch_client(config)
            raw = await client.indices.get_mapping(index=index_name)
            mappings = raw.get(index_name, {}).get("mappings", {})
            return {"index": index_name, "mappings": mappings}
        except Exception as e:
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.close()

    @app.get("/healthz/elasticsearch/indices/{index_name}/documents")
    async def elasticsearch_index_documents(
        index_name: str,
        request: Request,
        from_: int = Query(0, alias="from"),
        size: int = 20,
    ):
        from_val = max(0, from_)
        size_val = min(100, max(1, size))

        client = None
        try:
            config = ElasticsearchConfig()
            client = await get_elasticsearch_client(config)
            raw = await client.search(
                index=index_name,
                from_=from_val,
                size=size_val,
                body={"query": {"match_all": {}}},
            )

            total_hits = raw["hits"]["total"]
            total = total_hits if isinstance(total_hits, int) else total_hits.get("value", 0)

            documents = [
                {"_id": hit["_id"], "_source": hit["_source"]}
                for hit in raw["hits"]["hits"]
            ]

            return {
                "index": index_name,
                "total": total,
                "from": from_val,
                "size": size_val,
                "documents": documents,
            }
        except Exception as e:
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.close()

File: platform/fastify_server/routes/healthz_elasticsearch_data.route.mjs
import {
  ElasticsearchConfig,
  getSyncElasticsearchClient,
} from "@internal/db-connection-elasticsearch";

/**
 * Mount elasticsearch data-exploration routes.
 * These are read-only endpoints for browsing indices, documents, and field mappings.
 * @param {import('fastify').FastifyInstance} server
 */
export async function mount(server) {
  // Helper: create a short-lived client
  async function withClient(fn) {
    const config = new ElasticsearchConfig();
    const client = getSyncElasticsearchClient(config);
    try {
      return await fn(client, config);
    } finally {
      await client.close();
    }
  }

  // Normalise response body across ES v8 (returns body directly) and OpenSearch (returns { body })
  function body(response) {
    return response && typeof response === "object" && "body" in response && "statusCode" in response
      ? response.body
      : response;
  }

  // ── GET /healthz/elasticsearch/indices ──
  server.get("/healthz/elasticsearch/indices", async (_request, reply) => {
    try {
      return await withClient(async (client, config) => {
        const raw = await client.cat.indices({ format: "json" });
        const rows = body(raw);

        const indices = rows.map((r) => ({
          index: r.index,
          health: r.health,
          status: r.status,
          docsCount: r["docs.count"],
          storeSize: r["store.size"],
        }));

        return {
          configuredIndex: config.options.index || null,
          indices,
        };
      });
    } catch (err) {
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });

  // ── GET /healthz/elasticsearch/indices/:index/mappings ──
  server.get("/healthz/elasticsearch/indices/:index/mappings", async (request, reply) => {
    const { index } = request.params;
    try {
      return await withClient(async (client) => {
        const raw = await client.indices.getMapping({ index });
        const data = body(raw);
        const mappings = data[index]?.mappings || {};
        return { index, mappings };
      });
    } catch (err) {
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });

  // ── GET /healthz/elasticsearch/indices/:index/documents?from=0&size=20 ──
  server.get("/healthz/elasticsearch/indices/:index/documents", async (request, reply) => {
    const { index } = request.params;
    const from = Math.max(0, parseInt(request.query.from, 10) || 0);
    const size = Math.min(100, Math.max(1, parseInt(request.query.size, 10) || 20));

    try {
      return await withClient(async (client) => {
        const raw = await client.search({
          index,
          from,
          size,
          body: { query: { match_all: {} } },
        });
        const data = body(raw);

        const total =
          typeof data.hits.total === "number"
            ? data.hits.total
            : data.hits.total?.value ?? 0;

        const documents = data.hits.hits.map((hit) => ({
          _id: hit._id,
          _source: hit._source,
        }));

        return { index, total, from, size, documents };
      });
    } catch (err) {
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });
}
